### PR TITLE
[Mobile] - Add E2E tests for the Drag & Drop blocks feature

### LINKED ISSUE #4882: [Drag & Drop blocks] Add E2E tests
Add E2E tests related to the Drag & drop blocks feature, depending on which makes sense to add some might be left out to be done in manual testing.

### PR DESCRIPTION
**Pull request by geriux**
Closes https://github.com/wordpress-mobile/gutenberg-mobile/issues/4882

- `Gutenberg Mobile PR` -> https://github.com/wordpress-mobile/gutenberg-mobile/pull/4901

## What?
This PR adds some basic E2E tests for the Drag & Drop blocks functionality, mostly focused on the interaction with text-based blocks and long-pressing events.

## Why?
Since some parts of the Drag & drop blocks functionality rely on the native implementation and the behavior between iOS/Android and Aztec, some E2E tests are needed to cover these cases.

## How?

It adds the following tests:

- Drag & drop a block (Spacer block)
- Long-pressing on a text-based block (Paragraph), focusing on the text input, and pasting some text from the clipboard.
- Long-pressing on a text-based block that uses the `PlainText` component, focusing on the text input and pasting some text from the clipboard.
- Drag & drop a text-based block when the text input is not focused.

This PR also adds two new helpers related to the clipboard to use as wrappers to simplify their usage:

- setClipboard
- clearClipboard

## Testing Instructions

Check the full E2E tests pass in the [Gutenberg mobile PR](https://github.com/wordpress-mobile/gutenberg-mobile/pull/4901).

**Note**: There is some flakiness happening with the E2E tests so let's focus on the Drag & drop tests to pass.

## Screenshots or screencast 

https://user-images.githubusercontent.com/4885740/171022462-83a4e8da-1993-4863-a83c-e72a97dc8ee6.mov

https://user-images.githubusercontent.com/4885740/171022525-5e727cb8-069f-411b-81a8-77490f8141a2.mov

### CODE REVIEW

File: packages/react-native-editor/__device-tests__/gutenberg-editor-paste.test.js
Comment (geriux) [inline packages/react-native-editor/__device-tests__/gutenberg-editor-paste.test.js:61]: I moved this to `tapPasteAboveElement` so it handles all of the logic.

File: packages/react-native-editor/__device-tests__/helpers/utils.js
Comment (geriux) [inline packages/react-native-editor/__device-tests__/helpers/utils.js:325]: I changed this from `press` to `longPress` since it was causing random fails on iOS.

File: packages/react-native-editor/__device-tests__/helpers/utils.js
Comment (geriux) [inline packages/react-native-editor/__device-tests__/helpers/utils.js:359]: `tapPasteAboveElement` was only being used on Android, so I added support for iOS.

File: packages/react-native-editor/__device-tests__/helpers/utils.js
Comment (geriux) [inline packages/react-native-editor/__device-tests__/helpers/utils.js:452]: I ended up using `press` since `longPress` was not triggering correctly for the Drag & Drop feature on iOS.

File: packages/react-native-editor/__device-tests__/pages/editor-page.js
Comment (geriux) [inline packages/react-native-editor/__device-tests__/pages/editor-page.js:502]: This was needed to only get the `View` wrapper and no the TextInput

File: packages/react-native-editor/__device-tests__/gutenberg-editor-drag-and-drop.test.js
Comment (fluiddot) [inline packages/react-native-editor/__device-tests__/gutenberg-editor-drag-and-drop.test.js]: Typo:
```suggestion
		// Set clipboard text
```

File: packages/react-native-editor/__device-tests__/gutenberg-editor-drag-and-drop.test.js
Comment (fluiddot) [inline packages/react-native-editor/__device-tests__/gutenberg-editor-drag-and-drop.test.js]: I'm wondering if we could expand the `setClipboard` helper to support strings to prevent having to convert them into base64.

**Review by fluiddot — COMMENTED** [review]
I tested the E2E tests locally and found out that on iOS, the last one (i.e. `should be able to drag & drop a text-based block when the textinput is not focused`) gets stuck when trying to set the HTML. The error I'm encountering is:
```
"//XCUIElementTypeButton[@name="Switch to HTML"]" is still not visible after 25 retries!
```

As far as I checked, this issue is most likely related to an overlap of the system notification related to the copy/paste and the `...` button. In fact, I noticed that the `...` button is actually found but fails when clicking.

Probably we could address it by clearing the clipboard once is no longer required, wdyt?



_Screenshot took at the beginning of test case `should be able to drag & drop a text-based block when the textinput is not focused`._

**Comment by geriux** [comment]
> As far as I checked, this issue is most likely related to an overlap of the system notification related to the copy/paste and the `...` button. In fact, I noticed that the `...` button is actually found but fails when clicking.
> 
> Probably we could address it by clearing the clipboard once is no longer required, wdyt?

Thanks for spotting that! I didn't encounter the issue when I tested it locally. I tried setting the clipboard to an empty string but that does not clear the system notification, I'm thinking of increasing a bit the timer we have [here](https://github.com/WordPress/gutenberg/blob/trunk/packages/react-native-editor/__device-tests__/helpers/utils.js#L435-L436), what do you think?

Unfortunately, there's no option to clear the clipboard from the Appium device's API 😞

File: packages/react-native-editor/__device-tests__/gutenberg-editor-drag-and-drop.test.js
Comment (geriux) [reply packages/react-native-editor/__device-tests__/gutenberg-editor-drag-and-drop.test.js]: This is calling the `setClipboard` API from [Appium](https://appium.io/docs/en/commands/device/clipboard/set-clipboard/). Maybe we could create a helper to have a wrapper around it with this logic.

File: packages/react-native-editor/__device-tests__/gutenberg-editor-drag-and-drop.test.js
Comment (geriux) [reply packages/react-native-editor/__device-tests__/gutenberg-editor-drag-and-drop.test.js]: Oops! I'll update it! Thanks 👍

File: packages/react-native-editor/__device-tests__/gutenberg-editor-drag-and-drop.test.js
Comment (fluiddot) [reply packages/react-native-editor/__device-tests__/gutenberg-editor-drag-and-drop.test.js]: Yeah, good idea, that's exactly what I was thinking of 👍 .

Comment (geriux) [reply packages/react-native-editor/__device-tests__/gutenberg-editor-drag-and-drop.test.js]: Cool I'll include it in this PR since it's only used in a few places.

**Comment by fluiddot** [comment]
> Thanks for spotting that! I didn't encounter the issue when I tested it locally. I tried setting the clipboard to an empty string but that does not clear the system notification, I'm thinking of increasing a bit the timer we have [here](https://github.com/WordPress/gutenberg/blob/trunk/packages/react-native-editor/__device-tests__/helpers/utils.js#L435-L436), what do you think?
> 
> Unfortunately, there's no option to clear the clipboard from the Appium device's API 😞

I'm running the tests again and surprisingly I'm not getting the notification that overlaps the button 🤔. Probably, since I have the pasteboard synced with the simulator, I might have done some copy/pasting during the run and interfered with the tests 😅 . After doing more runs, I confirm that the E2E tests pass on iOS ✅ .

File: packages/react-native-editor/__device-tests__/pages/editor-page.js
Comment (fluiddot) [inline packages/react-native-editor/__device-tests__/pages/editor-page.js:806]: As far as I tested locally, this locator can't be found when running the Android E2E tests locally. I re-run the tests several times and always fail when querying this element, @geriux have you encountered this issue too? 

**Error message:**
```
"//android.view.ViewGroup[@content-desc="Shortcode Block. Row 1"]/android.view.ViewGroup/android.view.ViewGroup/android.widget.EditText" is still not visible after 25 retries!
```

Comment (fluiddot) [reply packages/react-native-editor/__device-tests__/pages/editor-page.js:806]: I checked the XPath of the TextInput element of the Shortcode using the Appium inspector, and contains a small difference (note that `android.view.ViewGroup` has `[2]` at the end.):
```
//android.view.ViewGroup[@content-desc="Shortcode Block. Row 1"]/android.view.ViewGroup[2]/android.widget.EditText
```

In fact, I tried to use the path we're using in the test (i.e. `//android.view.ViewGroup[@content-desc="Shortcode Block. Row 1"]/android.view.ViewGroup/android.view.ViewGroup/android.widget.EditText`) and the Appium inspector can't find the element 🤔 .

Comment (geriux) [reply packages/react-native-editor/__device-tests__/pages/editor-page.js:806]: Interesting, just to double-check, do you have `English` set as the device language?

Comment (geriux) [reply packages/react-native-editor/__device-tests__/pages/editor-page.js:806]: Sorry, just saw your next message with the different locator, I'll investigate this.

Comment (fluiddot) [reply packages/react-native-editor/__device-tests__/pages/editor-page.js:806]: > Interesting, just to double-check, do you have English set as the device language?

Yeah, the device language is English.

> Sorry, just saw your next message with the different locator, I'll investigate this.

No worries 😊. I see that the E2E tests passed in the GB-mobile PR, so this might be an issue related to the emulator, I'm using `Pixel 3 (Android 9)` in case it helps.

Comment (geriux) [reply packages/react-native-editor/__device-tests__/pages/editor-page.js:806]: So I'm trying this locally and I get no errors, I can look for the element using the Appium inspector successfully searching for the `XPath`: 

```
//android.view.ViewGroup[@content-desc="Shortcode Block. Row 1"]/android.view.ViewGroup/android.view.ViewGroup/android.widget.EditText
```

I wonder why it doesn't work for you 🤔 

If I manually inspect the element I do see the XPath you mentioned above:

```
//android.view.ViewGroup[@content-desc="Shortcode Block. Row 1"]/android.view.ViewGroup[2]/android.view.ViewGroup/android.widget.EditText
```

Comment (geriux) [reply packages/react-native-editor/__device-tests__/pages/editor-page.js:806]: >I'm using Pixel 3 (Android 9) in case it helps.

I'll try that one

Comment (fluiddot) [reply packages/react-native-editor/__device-tests__/pages/editor-page.js:806]: > So I'm trying this locally and I get no errors, I can look for the element using the Appium inspector successfully searching for the `XPath`:
> 
> ```
> //android.view.ViewGroup[@content-desc="Shortcode Block. Row 1"]/android.view.ViewGroup/android.view.ViewGroup/android.widget.EditText
> ```
> 
> I wonder why it doesn't work for you 🤔
> 
> If I manually inspect the element I do see the XPath you mentioned above:
> 
> ```
> //android.view.ViewGroup[@content-desc="Shortcode Block. Row 1"]/android.view.ViewGroup[2]/android.view.ViewGroup/android.widget.EditText
> ```

Interesting, I see that the path is different for me. In mine, `EditText` has only one `ViewGroup` element and yours have two 🤔 :

The Xpath I'm getting:
```
//android.view.ViewGroup[@content-desc="Shortcode Block. Row 1"]/android.view.ViewGroup[2]/android.widget.EditText
```

Your Xpath:
```
//android.view.ViewGroup[@content-desc="Shortcode Block. Row 1"]/android.view.ViewGroup[2]/android.view.ViewGroup/android.widget.EditText
```

Comment (fluiddot) [reply packages/react-native-editor/__device-tests__/pages/editor-page.js:806]: In any case, if the CI jobs of the E2E tests are passing I wouldn't block the PR due to this issue.

**Review by fluiddot — APPROVED** [review]
LGTM 🎊 ! Awesome job @geriux on adding the E2E tests, it was amazing watching the simulator/emulator dragging and dropping blocks automatically 🤩 .

Regarding the issue outlined in [this comment](https://github.com/WordPress/gutenberg/pull/41368#discussion_r885655884), since it's only happening locally on my environment and succeeding on CI jobs, we could skip it.

**Comment by geriux** [comment]
> Regarding the issue outlined in [this comment](https://github.com/WordPress/gutenberg/pull/41368#discussion_r885655884), since it's only happening locally on my environment and succeeding on CI jobs, we could skip it.

The full [tests passed](https://app.circleci.com/pipelines/github/wordpress-mobile/gutenberg-mobile/17644/workflows/a4182d69-4777-4de9-9fbd-5563de5d1cd3) 🎉 

I wonder if it was related to an outdated bundle or something. Either way, I'll keep an eye out if there are any failings after it is merged.

Thank you for reviewing! 🙇

**Comment by fluiddot** [comment]
> I wonder if it was related to an outdated bundle or something. Either way, I'll keep an eye out if there are any failings after it is merged.

Yep, most likely it's caused by that because I've just reinstalled the app in the emulator and now I'm getting the infamous error 😅 :
```
java.lang.ClassCastException: java.lang.Double cannot be cast to java.lang.Boolean
```

**Comment by fluiddot** [comment]
> > I wonder if it was related to an outdated bundle or something. Either way, I'll keep an eye out if there are any failings after it is merged.
> 
> Yep, most likely it's caused by that because I've just reinstalled the app in the emulator and now I'm getting the infamous error 😅 :
> 
> ```
> java.lang.ClassCastException: java.lang.Double cannot be cast to java.lang.Boolean
> ```

Finally, I've managed to address the issue I was having, and I can confirm that it was caused by an outdated version. Now I'm getting the same Xpath you shared 🎊 .
```
//android.view.ViewGroup[@content-desc="Shortcode Block. Row 1"]/android.view.ViewGroup[2]/android.view.ViewGroup/android.widget.EditText
```

**Comment by geriux** [comment]
> Finally, I've managed to address the issue I was having, and I can confirm that it was caused by an outdated version. Now I'm getting the same Xpath you shared 🎊 .

Yay! Thanks for double-checking 🎉

**Pull request merged**
